feat: use DAISY Pipeline app if available

NPavie opened the pull request:
This PR changes the initialisation process of the command line interface, to allow the use of the DAISY Pipeline electron app :

Additionally, the following settings have been removed or renamed as requested :
- `exec_line` has been renamed as `app_path`
  - If this setting is pointing to a DAISY Pipeline app executable (`DAISY Pipeline` or `DAISY Pipeline.exe`), it will try to launch the app with the requested command line arguments
  - Using the string value `DAISY Pipeline` will look for the DAISY Pipeline app in the user path.
- `ws_timeup` has been removed
  - It's usage in code is replaced by the previous default value `25` 
- `local` has been removed
- `timeout` has been removed
  - It's usage in code is replaced by the previous default value `25`

bertfrees commented:
@NPavie Would it be possible to allow omitting the `--output` option when going through the app, and let the app use its result folder (which can be set in the settings) when the output option was omitted?

NPavie commented:
Quick resume of DAISY meeting regarding this :
- `--output` omission should be feasible on the pipeline-ui side (parameters are not checked before the args forwarding to the app
- I'll add an improved config file checking process, following this order of config file presence : 
  - Check the commande line flag (with -f or --file)
  - Else from the current working directory
  - Else next to the dp2 tool
  - Else use a default config that tries to looks for the app near the dp2 tool or in the path

bertfrees commented on cli/cli.go:
```diff
@@ -18,8 +18,7 @@ const (
 const (
-	MAIN_HELP_TEMPLATE = `
-Usage {{.Name}} [GLOBAL_OPTIONS] command [COMMAND_OPTIONS] [PARAMS]
+	MAIN_HELP_TEMPLATE = `Usage {{.Name}} [GLOBAL_OPTIONS] command [COMMAND_OPTIONS] [PARAMS]
```
Does this mean the empty line before "Usage dp2 ..." is removed? Could this be put in a separate commit?

NPavie replied:
Oups might have wrongly added this to the PR

(Got mixed up with the vs code go extension doing auto format with goeverywhere)

I might clean the PR branch to discard the reformat stuff that got in the way.

bertfrees commented on cli/cli.go:
```diff
@@ -84,18 +80,17 @@ List of global options:       {{.Parent.Name}} help -g
 `
 
-//{{range .Flags}}{{if len .Short}}-{{.Short}},{{end}}--{{.Long}}{{if isOption .}}   {{upper .Long}}{{end}}       {{.Description}}{{end}}
+// {{range .Flags}}{{if len .Short}}-{{.Short}},{{end}}--{{.Long}}{{if isOption .}}   {{upper .Long}}{{end}}       {{.Description}}{{end}}
```
All this reformatting is good (or most of it), but it would be nice if it were in a separate commit, so we could better see what actually changed (namely the `bringUp` function).

bertfrees replied:
(Same remark for the "chore: cleanup" commit.)

NPavie replied:
(Found a way to deactivate format on save enforced by my vscode Go extension, on the way to ditch it)

bertfrees commented on cli/config.go:
```diff
@@ -77,18 +77,44 @@ func NewConfig() Config {
+		cnf = copyConf()
+		if cnf[DEBUG].(bool) || slices.Contains(os.Args[1:], "--debug") {
+			log.Println(err.Error())
+		}
 	}
+	cnf.UpdateDebug()
```
I would even go further and print the "Warning : no default configuration file found" only when debugging is enabled.

Oh and one more thing: shouldn't the `slices.Contains(os.Args[1:], "--debug")` check for a value `true`?

NPavie replied:
Oh wait it's expecting true or false value, i'll change that

bertfrees commented on cli/link.go:
```diff
@@ -84,17 +86,47 @@ func (p PipelineLink) IsLocal() bool {
+		// Requested behavior :
+		// dp2 launches pipeline-ui (if needed); dp2 calls pipeline-ui as cli handler; pipeline-ui calls dp2 with the port info as an argument
+
+		// When port attribute is set, the cli tool should not try to start a pipeline instance
+		// and directly connect to pipeline instance for which the port is set on the command line
+		if pLink.config[STARTING].(bool) && !(slices.Contains(os.Args[1:], "--port") || slices.Contains(os.Args[1:], "--host")) {
```
I think we should not only check the command, but also check in the config file whether `port`, `host` (or `ws_path`) has been set.

NPavie replied:
so the default config for those field would be set to empty right ?

bertfrees replied:
In the config.yml file of the app, these settings would be absent.

They could possibly still be defined in the default config in the Go code, however it's important that these default values are only used if `app_path` is not set or the app can not be found (or possibly also if the app is not running and `starting` is false). Otherwise, the CLI embedded in the app will not be able to connect to the app while a server is running on localhost:8080.

The benefit of keeping some default, is that `dp2` will still "just work" even without a config file, and that's a nice thing to have. I think it would be nice if the defaults defined in the code match the general config.yml, so that config.yml is only really needed for users that want to edit it (to use as template), and for embedment in the app.

NPavie replied:
> (or possibly also if the app is not running and starting is false)

This implies that the code have to check if the app is running depending on the "starting" parameter, right ?
(It should be doable but might need a little research, possibly using a package Marisa mentionned)

So if I summary, forwarding args to the app should only be done if :
```
No host/port/wspath defined in config (including command line flags)
and (
    app path is set 
    and app exists 
    and (
        starting is true
        or app is already started
    )
) 
```
And with that, we assume the custom config file distributed with the app would have the following fields set
- host/port/wspath undefined or set to empty
- app_path set to the app it is embedded with
- starting set to true

bertfrees replied:
>> (or possibly also if the app is not running and starting is false)
>
> This implies that the code have to check if the app is running depending on the "starting" parameter, right?

Hmm, did not think of that, actually.

But indeed, strictly speaking if we want to support `starting=false`, we need such an additional check!

> And with that, we assume the custom config file distributed with the app would have the following fields set
> -  starting set to true

Yes, I guess it is fine to set it to true. The alternative is to set it to false, and show an informative message that the app must be started manually, or that the `--starting` switch must be passed (instead of the generic "Could not connect to the webservice and I'm not configured to start one" message we currently have). In addition, there could be a setting in the app to enable/disable the automatic starting behavior.

bertfrees commented on cli/link.go:
```diff
@@ -84,17 +86,47 @@ func (p PipelineLink) IsLocal() bool {
+				} else {
+					os.Exit(cmd.ProcessState.ExitCode())
+				}
+			} else {
+				alive, err = NewPipelineLauncher(
+					pLink.pipeline,
+					execpath,
+					10,
+				).Launch(os.Stdout)
+				if err != nil {
+					return fmt.Errorf("Error bringing the pipeline2 up %v", err.Error())
+				}
 			}
```
I didn't think about retaining the support for other web service launcher scripts. I guess it is indeed a possibility. I'm not totally convinced of the usefulness of this feature, now that the web service is launched through the app, but it doesn't hurt either. Two comments:

- If the executable is not the app, it should be allow to have `port`/`host`/`ws_path` in the config/command.
- I'm not sure about reusing the `app_path` setting. OTOH having two settings (`app_path` and `exec_line`) is also a bit confusing.

bertfrees commented on cli/link.go:
```diff
@@ -84,17 +86,47 @@ func (p PipelineLink) IsLocal() bool {
+				cmd := exec.Command(absPath, args...)
+				cmd.Stdout = os.Stdout
+				cmd.Stderr = os.Stderr
+				if err := cmd.Run(); err != nil {
+					return fmt.Errorf("Could not use the pipeline app: %v", err)
+				} else {
+					os.Exit(cmd.ProcessState.ExitCode())
+				}
```
Is there nothing like the [`exec()`](https://man7.org/linux/man-pages/man3/exec.3.html) function in Go?

NPavie replied:
Hum not sure

(Note that I don't think exec has an equivalent on windows to "replace" an ongoing executable process, so maybe they cannot make a replicate of that in go)

bertfrees replied:
Ah yes, that could be the reason.

bertfrees commented on dp2/config.yml:
```diff
@@ -1,17 +1,12 @@
-timeout: 10
 #debug
 debug: false
 starting: true
```
Should we change `starting` to false? In the config.yml that we ship with the app, it should of course be set to `true`, but I don't know if there is a need/reason to set it to true in other cases. The same goes for `app_path`: In the config.yml that we ship with the app, this field should be set to the relative path to the app executable, but in the general case I think we can assume there is no app?

NPavie replied:
Yeah we might need to assume for general distribution that app is not available.

I'm not really sure what would be the default config file to be shipped for general distribution.

bertfrees replied:
IMO it would make sense to have the same values in the file as in the Go code, and with the same comments as in the help command.

```yaml
# Pipeline's webservice host
host: http://localhost
# Pipeline's webserivce port
port: 8181
# Pipeline's webservice path, as in HOST:PORT/PATH (e.g. http://localhost:8181/ws)
path: ws
# Path of the DAISY Pipeline app or the pipeline launch script to use the webservice.
# If empty or only containing 'DAISY Pipeline', the app is searched in the PATH
#app_path:
# Client key for authenticated requests
#client_key:
# Client secrect for authenticated requests
#client_secret:
# Print debug messages
debug: false
# Start the webservice or the DAISY Pipeline app in the local computer if it is not running
starting: false
```

Note that all settings could have been commented out (because they're the default anyway), but I chose to comment out only the ones with an empty/no value, to make it slightly clearer. We could possibly change the values of those commented out lines, so they become suggestions for possible values:

```yaml
# Client key for authenticated requests
#client_key: your-id
# Client secrect for authenticated requests
#client_secret: your-password
```

bertfrees commented on dp2/config.yml:
```diff
@@ -1,17 +1,12 @@
-#WS CONFIGURATION
+#WS CONFIGURATION if app is not available
```
What does "if the app is not available" mean here?

NPavie replied:
I was thinking to use the WS configuration if app is not available in path or through the app_path variable.

(but from your comment in the argument test to launch the app, that would need to be change)

bertfrees commented on cli/link.go:
```diff
@@ -84,17 +86,43 @@ func (p PipelineLink) IsLocal() bool {
 	if err != nil {
-		if pLink.config[STARTING].(bool) {
-			alive, err = NewPipelineLauncher(pLink.pipeline,
-				pLink.config.ExecPath(), pLink.config[TIMEOUT].(int)).Launch(os.Stdout)
```
> timeout has been removed
>
> It's usage in code is replaced by the previous default value 25

If we are keeping the support for launch scripts, it might make sense to also keep the timeout setting.

bertfrees commented:
> ws_timeup has been removed
>
> It's usage in code is replaced by the previous default value 25

This setting was not used anymore, or did I look wrong?

NPavie commented:
@bertfrees I reseted the branch to take in account your latest remarks and behaviour requests, and added some additionnal changes.

I removed all the reformat that was done automatically by my vscode extension.

On the behavior change

-  in the config side, if app_path is either not set or set to an invalid path
    - webservice settings are set with the default values declared in code
    - app_path is reset to empty
    - starting is reset to false

I also added a small sanitizing in the config url function to allow the use of slash at the begining or the end of the webservice path in the config.
(So if pipeline webservice is to be bound to the root of an address like http://localhost;8181/  on a custom install of the pipeline, users can still point to it, it won't be discarded in default configuration evaluation and will be resolved as previously)

- In the link side :
    - if web service fields are set (either by the user or by reloading the default values due to invalid/empty app_path), the previous behaviour is kept
    - else if pipeline app is found running on the system, commands are forward to the app
    - else if starting is set to true, the tool tries to launch the app with the commands
    - else the default webservice is tried
    - else the error stating the app nor the webservice could be used

I added a dependency to the gp-ps package to list running process on the user system.

bertfrees commented:
@NPavie Sounds good, thanks for the changes. I'll look at it in more detail.

NPavie commented:
Oh wait, my sanitizing in the url function is not correct (i did not reread the format string)
Will change that asap

bertfrees commented on cli/config.go:
```diff
@@ -141,7 +209,27 @@ func (c Config) ExecPath() string {
 func (c Config) buildPath(base string) string {
-	p := filepath.FromSlash(c[EXECLINE].(string))
+	execpath := c[APPPATH].(string)
+	//empty app path defaults to looking for DAISY Pipeline app in the PATH
```
Notes to self (possibly add these as comments):
- We choose not to use "DAISY Pipeline" as the default config value, because this would result in a "provided app path was not found" warning if no app is found.
- Having no default value at all would also have been a possibility, because the field will be set in the config.yml that we ship with the app anyway, and in the general case we can assume there is no app. However it doesn't hurt to add it either, especially because the app will not be searched for anyway when `host`, `port` and `ws_path` are set.

bertfrees commented on cli/config_test.go:
```diff
@@ -15,8 +15,7 @@ var (
 port: 9999
-ws_path: ws
-ws_timeup: 10
+path: ws
```
I kinda liked `ws_path`, rather than `path`. Especially now that we have another "path", namely `app_path`.

NPavie replied:
We can change it back as you like.

bertfrees commented on dp2/config.yml:
```diff
@@ -1,18 +1,20 @@
-#WS CONFIGURATION
+# Pipeline's webservice host
 host: http://localhost
+# Pipeline's webservice port
 port: 8181
```
A note saying that these should be (left commented out if one wishes to use `app_path`, would be useful.

bertfrees commented on dp2/config.yml:
```diff
@@ -1,18 +1,20 @@
-#debug
-debug: false
-starting: true
+# Path of the DAISY Pipeline app or the pipeline launch script to use the webservice.
+# If empty or only containing 'DAISY Pipeline', the app is searched in the PATH
```
Are paths resolved relative to this config file? If so, it's probably good to add a note here and in the help.

NPavie replied:
Hi @bertfrees, I rechecked the code that resolves the path : if it is a relative path that is not registered in the PATH variable (that's the bit I added in the path retrieval), it is evaluated as relative to the "executable folder" in the config.go functions that resolve it.

From the doc of the function, this is the folder that contains the running executable (but i'm not sure if this resolve symlinked executables)

https://github.com/NPavie/pipeline-cli-go/blob/88ce4298e1afb356ec61640ba97b013f2cf703ab/cli/config.go#L219-L242

NPavie replied:
Thinking about it, it might be a good idea to change this logic and indeed use a "config relative" evaluation of the path if the path is not absolute or if it is not an executable exposed in the PATH environment variable

bertfrees replied:
Yes, I too think that makes sense.

bertfrees replied:
@NPavie Will you make that change?

NPavie replied:
i'm on it

bertfrees replied:
thanks!

bertfrees commented:
@NPavie I have rebased the branch and made some additions.

NPavie commented:
@bertfrees there is a test reported in error in the build, that is passing on my dev environment with my latest changes.
(If i read the log correctly, it's the test where a config file is passed as an argument of the command line. In the test, the config file path seems not correctly kept in memory, maybe i missed something for non-windows system)

Can you test it on your side ?

bertfrees commented on cli/config.go:
```diff
@@ -155,11 +162,30 @@ func (c Config) FromYaml(r io.Reader) error {
+		apppath := c.AppPath()
+		switch runtime.GOOS {
+		case "windows":
+			for _, ext := range WIN_EXT {
+				if _, err := os.Stat(apppath + ext); !errors.Is(err, os.ErrNotExist) {
+					// found an executable with the given extension
+					c[APPPATH] = apppath + ext
+					break
+				}
+			}
```
@NPavie Just checking: isn't this check already done within `AppPath()`?

NPavie replied:
(will do it to while i'm at it just in case)

bertfrees commented:
>  In the test, the config file path seems not correctly kept in memory

This was because `cli.go` is generated from `cli.go.in`, so you need to edit `cli.go.in`.

NPavie commented:
oh ok ! thanks for the fix !

bertfrees commented:
OK I'm going to make release now.
